### PR TITLE
[F] Simplify  cache adapter interface

### PR DESCRIPTION
**Pull request by Istar-Eldritch**
Pending tasks:
-  [x] Tests

### CODE REVIEW

**Comment by Istar-Eldritch** [comment]
This is ready to go.

File: event-store/src/adapters/pg/mod.rs [hunk 1]
```diff
@@ -22,11 +24,35 @@ pub struct PgQuery<'a> {
...
 
-impl<'a> StoreQuery for PgQuery<'a> {}
+impl<'a> StoreQuery for PgQuery<'a> {
+    fn unique_id(&self) -> String {
+        let hash = Sha256::digest(format!("{:?}:[{}]", self.args, self.query).as_bytes());
+        let result = hash.iter().fold(String::new(), |mut acc, hex| {
+            acc.push_str(&format!("{:X}", hex));
+            acc
+        });
+        result
```
Comment (jamwaffles) [inline event-store/src/adapters/pg/mod.rs]: You can get rid of the redundant `result` variable here.

File: event-store/src/aggregator.rs [hunk 2]
```diff
@@ -72,9 +72,14 @@ use store_query::StoreQuery;
...
 ///     }
 /// }
-/// #
-/// # struct DummyQuery;
-/// # impl StoreQuery for DummyQuery {}
+///
+/// struct DummyQuery;
+/// impl StoreQuery for DummyQuery {
+///   fn unique_id(&self) -> String {
+///     // Please read docs about unique_id
```
Comment (jamwaffles) [inline event-store/src/aggregator.rs]: You should change this to a note about generating a unique identifier from the contents of the query and its args here; it's vitally important for the cache to work correctly.

File: event-store/src/lib.rs [hunk 3]
```diff
@@ -93,32 +93,36 @@ where
     }
 
     /// Query the backing store and return an entity `T`, reduced from queried events
-    fn aggregate<'b, E, T, A>(&self, query_args: A) -> BoxedFuture<'b, T, String>
+    fn aggregate<'b, E, T, A>(&'b self, query_args: A) -> BoxedFuture<'b, T, String>
```
Comment (jamwaffles) [inline event-store/src/lib.rs]: Can you see if you can get rid of this lifetime? I feel like putting explicit lifetimes on `self` is a big code smell. I could be wrong and it's required, but it'd be killa if it wasn't.

File: event-store/src/aggregator.rs (hunk 2 above)
Comment (jamwaffles) [reply event-store/src/aggregator.rs]: It does depend on the impl, yes, but the desired outcome is the same; a unique cache key for the given query/args, in whatever format they might be in. I think it's worth at least clarifying this comment a bit more.

File: event-store/src/lib.rs (hunk 3 above)
Comment (jamwaffles) [reply event-store/src/lib.rs]: Addendum: there's [this SO question](https://stackoverflow.com/questions/31641610/return-reference-with-lifetime-of-self) which seems to suggest `'b` can be elided. Would that work?

File: event-store/src/adapters/pg/cache.rs [hunk 4]
```diff
@@ -20,54 +21,58 @@ impl PgCacheAdapter {
...
+                ).expect("Retrieve cache");
 
-        // `rows.get()` panics if index is out of bounds, hence this check
-        if rows.len() != 1 {
-            None
-        } else {
-            let row = rows.get(0);
+            // `rows.get()` panics if index is out of bounds, hence this check
+            if rows.len() != 1 {
+                Box::new(FutOk(None))
```
Comment (petereast) [inline event-store/src/adapters/pg/cache.rs]: Does this need to be boxed? - it's inside `Box::new(LazyFuture(|| {...}))` so does it need to be boxed again?

Comment (Istar-Eldritch) [reply event-store/src/adapters/pg/cache.rs]: Read again

**Pull request merged**